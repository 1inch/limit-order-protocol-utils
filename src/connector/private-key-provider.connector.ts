import {ProviderConnector} from './provider.connector';
import Web3 from 'web3';
import {EIP712TypedData} from '../model/eip712.model';
import {AbiItem} from '../model/abi.model';
import {AbiItem as Web3AbiItem} from 'web3-utils';
import {MsgParams, signTypedData_v4} from 'eth-sig-util';

export class PrivateKeyProviderConnector implements ProviderConnector {
    constructor(
        private readonly privateKey: string,
        protected readonly web3Provider: Web3
    ) {}

    contractEncodeABI(
        abi: AbiItem[],
        address: string | null,
        methodName: string,
        methodParams: unknown[]
    ): string {
        const contract = new this.web3Provider.eth.Contract(
            abi as Web3AbiItem[],
            address === null ? undefined : address
        );

        return contract.methods[methodName](...methodParams).encodeABI();
    }

    signTypedData(
        _walletAddress: string,
        typedData: EIP712TypedData,
        /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
        _typedDataHash = ''
    ): Promise<string> {
        const result = signTypedData_v4(Buffer.from(this.privateKey, 'hex'), {
            data: typedData,
            /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
        } as MsgParams<any>);

        return Promise.resolve(result);
    }

    ethCall(contractAddress: string, callData: string): Promise<string> {
        return this.web3Provider.eth.call({
            to: contractAddress,
            data: callData,
        });
    }

    decodeABIParameter<T>(type: string, hex: string): T {
        return this.web3Provider.eth.abi.decodeParameter(type, hex) as T;
    }
}
