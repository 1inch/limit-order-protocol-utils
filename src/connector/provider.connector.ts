import {EIP712TypedData} from '../model/eip712.model';
import {AbiItem, AbiOutput} from '../model/abi.model';

export interface AbiDecodeResult {
    [key: string]: unknown,
}

export interface ProviderConnector {
    contractEncodeABI(
        abi: AbiItem[],
        address: string | null,
        methodName: string,
        methodParams: unknown[]
    ): string;

    signTypedData(
        walletAddress: string,
        typedData: EIP712TypedData,
        typedDataHash: string
    ): Promise<string>;

    ethCall(contractAddress: string, callData: string): Promise<string>;

    decodeABIParameter<T>(type: AbiOutput | string, hex: string): T;

    decodeABICallParameters(types: Array<AbiOutput | string>, callData: string): AbiDecodeResult;
}
