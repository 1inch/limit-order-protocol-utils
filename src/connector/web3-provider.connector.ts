import {ProviderConnector} from './provider.connector';
import Web3 from 'web3';
import {EIP712TypedData} from '../model/eip712.model';
import {AbiItem} from '../model/abi.model';
import {AbiItem as Web3AbiItem} from 'web3-utils';

export class Web3ProviderConnector implements ProviderConnector {
    constructor(protected readonly web3Provider: Web3) {}

    contractEncodeABI(
        abi: AbiItem[],
        address: string | null,
        methodName: string,
        methodParams: unknown[]
    ): string {
        const contract = new this.web3Provider.eth.Contract(
            abi as Web3AbiItem[],
            address === null ? undefined : address
        );

        return contract.methods[methodName](...methodParams).encodeABI();
    }

    signTypedData(
        walletAddress: string,
        typedData: EIP712TypedData,
        /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
        _typedDataHash: string
    ): Promise<string> {
        if (!this.web3Provider.currentProvider) {
            throw new Error('Web3 currentProvider is null');
        }

        return this.web3Provider.currentProvider.send('eth_signTypedData_v4', [
            walletAddress,
            JSON.stringify(typedData),
        ]);
    }

    ethCall(contractAddress: string, callData: string): Promise<string> {
        return this.web3Provider.eth.call({
            to: contractAddress,
            data: callData,
        });
    }

    decodeABIParameter<T>(type: string, hex: string): T {
        return this.web3Provider.eth.abi.decodeParameter(type, hex) as T;
    }
}
