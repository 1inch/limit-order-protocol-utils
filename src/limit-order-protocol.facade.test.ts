import {LimitOrderProtocolFacade} from './limit-order-protocol.facade';
import Web3 from 'web3';
import {
    LimitOrder,
    LimitOrderProtocolMethods,
} from './model/limit-order-protocol.model';
import {LimitOrderBuilder} from './limit-order.builder';
import {LimitOrderPredicateBuilder} from './limit-order-predicate.builder';
import {FakeProviderConnector} from '../test/fake-provider.connector';
import {FILL_ORDER_SNAPSHOT} from '../test/fill-order-sanpshot';
import {CANCEL_ORDER_SNAPSHOT} from '../test/cancel-order-snapshot';

describe('LimitOrderProtocolFacade - facade for Limit order protocol contract', () => {
    const contractAddress = '0xe3456f4ee65e745a44ec3bcb83d0f2529d1b84eb';
    const walletAddress = '0xfb3c7eb936cAA12B5A884d612393969A557d4307';
    const chainId = 56;

    const privateKey =
        '552be66668d14242eeeb0e84600f0946ddddc77777777c3761ea5906e9ddcccc';
    const web3 = new Web3('https://bsc-dataseed.binance.org');

    let providerConnector: FakeProviderConnector;
    let facade: LimitOrderProtocolFacade;
    let limitOrderPredicateBuilder: LimitOrderPredicateBuilder;
    let limitOrderBuilder: LimitOrderBuilder;

    function createOrderWithPredicate(predicate: string): LimitOrder {
        return limitOrderBuilder.buildOrder({
            makerAssetAddress: '0xe9e7cea3dedca5984780bafc599bd69add087d56',
            takerAssetAddress: '0x111111111117dc0aa78b770fa6a738034120c302',
            makerAddress: walletAddress,
            makerAmount: '1000000000000000000',
            takerAmount: '1000000000000000000',
            predicate,
        });
    }

    beforeEach(() => {
        providerConnector = new FakeProviderConnector(privateKey, web3);

        facade = new LimitOrderProtocolFacade(
            contractAddress,
            providerConnector
        );

        limitOrderPredicateBuilder = new LimitOrderPredicateBuilder(facade);

        limitOrderBuilder = new LimitOrderBuilder(
            contractAddress,
            chainId,
            providerConnector
        );
    });

    describe('fillOrder()', () => {
        it('Must create a call data for order filling', async () => {
            const timestamp = 1621040104;
            const timestampBelow = limitOrderPredicateBuilder.timestampBelow(
                timestamp
            );
            const order = createOrderWithPredicate(timestampBelow);

            order.salt = '1';

            const typedData = limitOrderBuilder.buildOrderTypedData(order);
            const signature = await limitOrderBuilder.buildOrderSignature(
                walletAddress,
                typedData
            );
            const makerAmount = '1000000000000000000';
            const takerAmount = '0';
            const thresholdAmount = '0';

            const callData = facade.fillOrder(
                order,
                signature,
                makerAmount,
                takerAmount,
                thresholdAmount
            );

            expect(callData).toBe(FILL_ORDER_SNAPSHOT);
        });
    });

    describe('cancelOrder()', () => {
        it('Must create a call data for order canceling', async () => {
            const timestamp = 1621040104;
            const timestampBelow = limitOrderPredicateBuilder.timestampBelow(
                timestamp
            );
            const order = createOrderWithPredicate(timestampBelow);

            order.salt = '1';

            const callData = facade.cancelOrder(order);

            expect(callData).toBe(CANCEL_ORDER_SNAPSHOT);
        });
    });

    describe('nonce()', () => {
        it('Return the nonce number of address (for real wallet address)', async () => {
            const nonce = await facade.nonce(
                '0xbbcf91605c18a9859c1d47abfeed5d2cca7097cf'
            );

            expect(nonce).toBeGreaterThan(1);
        });

        it('Return 0 when address never called advanceNonce (for contract address)', async () => {
            const nonce = await facade.nonce(contractAddress);

            expect(nonce).toBe(0);
        });
    });

    describe('advanceNonce()', () => {
        it('Must create a call data for advance nonce', async () => {
            const callData = await facade.advanceNonce(2);

            expect(callData).toBe(
                '0x72c244a80000000000000000000000000000000000000000000000000000000000000002'
            );
        });
    });

    describe('increaseNonce()', () => {
        it('Must create a call data for increase nonce', async () => {
            const callData = await facade.increaseNonce();

            expect(callData).toBe('0xc53a0292');
        });
    });

    describe('remaining()', () => {
        it('When order is never been touched, then must throw error', async () => {
            const order = createOrderWithPredicate('0x');
            const orderTypedData = limitOrderBuilder.buildOrderTypedData(order);
            const orderHash = limitOrderBuilder.buildOrderHash(orderTypedData);

            let error = null;

            try {
                await facade.remaining(orderHash);
            } catch (e) {
                error = e;
            }

            expect(error?.message?.includes('LOP: Unknown order')).toBe(true);
        });

        it('When order is partially filled, then must return remaining amount', async () => {
            // Order WBNB > 1INCH, filled for 3%
            const orderHash =
                '0x969ecf29e9b02b15e9640a6c741ea2db7984a00843485e0ed6aa4638e6860838';

            const remaining = await facade.remaining(orderHash);

            expect(remaining.toString()).toBe('2917591702620370');
        });

        it('When order is canceled, then must return zero', async () => {
            // Canceled order
            const orderHash =
                '0x44579da3014205b53cafe1badd3a4e8327d769560a0f38efd9840c402670312c';

            const remaining = await facade.remaining(orderHash);

            expect(remaining.toString()).toBe('0');
        });
    });

    describe('checkPredicate()', () => {
        it('When the order predicates are valid then return true', async () => {
            const timestamp = Math.floor(Date.now() / 1000) + 600;
            const timestampBelow = limitOrderPredicateBuilder.timestampBelow(
                timestamp
            ); // valid value
            const predicate = await limitOrderPredicateBuilder.and(
                timestampBelow
            );

            const order = createOrderWithPredicate(predicate);

            const result = await facade.checkPredicate(order);

            expect(result).toBe(true);
        });

        it('When the order predicates are NOT valid then return false', async () => {
            const timestampBelow = facade.getContractCallData(
                LimitOrderProtocolMethods.timestampBelow,
                [12000000] // must be 0x0000...
            );
            const predicate = await limitOrderPredicateBuilder.and(
                timestampBelow
            );

            const order = createOrderWithPredicate(predicate);

            const result = await facade.checkPredicate(order);

            expect(result).toBe(false);
        });

        it('When provider returns invalid value, then return false', async () => {
            const predicate = await limitOrderPredicateBuilder.eq(
                '1',
                walletAddress,
                '0x000'
            );

            const order = createOrderWithPredicate(predicate);

            console.log(
                'This is ok! Error "overflow (fault="overflow", operation="toNumber"..." must be shown!'
            );
            const result = await facade.checkPredicate(order);

            expect(result).toBe(false);
        });
    });

    describe('simulateCalls()', () => {
        it('When an order is invalid by the nonce predicate then must return false', async () => {
            const timestamp = Math.floor(Date.now() / 1000) + 100;
            const timestampBelow = limitOrderPredicateBuilder.timestampBelow(
                timestamp
            ); // valid value
            const nonce = await facade.nonce(walletAddress); // real nonce
            const nonceEquals = limitOrderPredicateBuilder.nonceEquals(
                walletAddress,
                nonce + 1
            ); // invalid value

            const predicate = await limitOrderPredicateBuilder.and(
                timestampBelow,
                nonceEquals
            );
            const order = createOrderWithPredicate(predicate);

            const tokens = [contractAddress, walletAddress];
            const data = [order.predicate, order.makerAssetData];

            const result = await facade.simulateCalls(tokens, data);

            expect(result).toBe(false);
        });

        it('When an order is invalid by the timestamp predicate then must return false', async () => {
            const timestamp = Math.floor(Date.now() / 1000) - 100000; // invalid value
            const timestampBelow = limitOrderPredicateBuilder.timestampBelow(
                timestamp
            ); // valid value
            const nonce = await facade.nonce(walletAddress); // real nonce
            const nonceEquals = limitOrderPredicateBuilder.nonceEquals(
                walletAddress,
                nonce
            ); // valid value

            const predicate = await limitOrderPredicateBuilder.and(
                timestampBelow,
                nonceEquals
            );
            const order = createOrderWithPredicate(predicate);

            const tokens = [contractAddress, walletAddress];
            const data = [order.predicate, order.makerAssetData];

            const result = await facade.simulateCalls(tokens, data);

            expect(result).toBe(false);
        });

        it('When an order is valid by all predicates then must return true', async () => {
            const timestamp = Math.floor(Date.now() / 1000) + 100; // valid value
            const timestampBelow = limitOrderPredicateBuilder.timestampBelow(
                timestamp
            ); // valid value
            const nonce = await facade.nonce(walletAddress); // real nonce
            const nonceEquals = limitOrderPredicateBuilder.nonceEquals(
                walletAddress,
                nonce
            ); // valid value

            const predicate = await limitOrderPredicateBuilder.and(
                timestampBelow,
                nonceEquals
            );
            const order = createOrderWithPredicate(predicate);

            const tokens = [contractAddress, walletAddress];
            const data = [order.predicate, order.makerAssetData];

            const result = await facade.simulateCalls(tokens, data);

            expect(result).toBe(true);
        });
    });

    it('domainSeparator() return domain separator (used for signing typed data by Ledger)', async () => {
        const result = await facade.domainSeparator();

        expect(result).toBe(
            '0xc01dba435cd14ffa0d760af4ffb49214bf6c739da021e8377adedcb4c4db47e8'
        );
    });

    it("parseSimulateTransferResponse() return null when response doesn't contain special prefix", () => {
        const input =
            '1111111111' +
            '000000000000000000000000000000000000000000000000000000000000002' +
            '000000000000000000000000000000000000000000000000000000000000000' +
            '124c4f503a20556e6b6e6f776e206f726465720000000000000000000000000000';

        const result = facade.parseSimulateTransferResponse(input);

        expect(result).toBe(null);
    });

    it("parseSimulateTransferError() return null when response doesn't contain special prefix", () => {
        const input = 'dddddddddd';
        const result = facade.parseSimulateTransferError(input);

        expect(result).toBe(null);
    });

    it("parseSimulateTransferError() return true when response doesn't contain any zero chars", () => {
        const input = 'CALL_RESULTS_11';
        const result = facade.parseSimulateTransferError(input);

        expect(result).toBe(true);
    });

    it('parseSimulateTransferError() return false when response contain zero chars', () => {
        const input = 'CALL_RESULTS_01';
        const result = facade.parseSimulateTransferError(input);

        expect(result).toBe(false);
    });
});
