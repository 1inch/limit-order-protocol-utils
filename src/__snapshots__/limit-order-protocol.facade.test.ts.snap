// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LimitOrderProtocolFacade - facade for Limit order protocol contract cancelLimitOrder() Must create a call data for order canceling 1`] = `"0x2d9a56f600000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000001000000000000000000000000e9e7cea3dedca5984780bafc599bd69add087d56000000000000000000000000111111111117dc0aa78b770fa6a738034120c302000000000000000000000000fb3c7eb936caa12b5a884d612393969a557d4307000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000de0b6b3a76400000000000000000000000000000000000000000000000000000de0b6b3a764000000000024000000240000002400000024000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000140000000000000000000000000000000000000000000000000000000000000002463592c2b00000000000000000000000000000000000000000000000000000000609f1be800000000000000000000000000000000000000000000000000000000"`;

exports[`LimitOrderProtocolFacade - facade for Limit order protocol contract fillLimitOrder() Must create a call data for order filling 1`] = `"0x62e238bb00000000000000000000000000000000000000000000000000000000000000c0000000000000000000000000000000000000000000000000000000000000026000000000000000000000000000000000000000000000000000000000000002e00000000000000000000000000000000000000000000000000de0b6b3a7640000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000e9e7cea3dedca5984780bafc599bd69add087d56000000000000000000000000111111111117dc0aa78b770fa6a738034120c302000000000000000000000000fb3c7eb936caa12b5a884d612393969a557d4307000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000de0b6b3a76400000000000000000000000000000000000000000000000000000de0b6b3a764000000000024000000240000002400000024000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000140000000000000000000000000000000000000000000000000000000000000002463592c2b00000000000000000000000000000000000000000000000000000000609f1be800000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000004179f0370195da42aef392ddc9d3254827146a8476c559fcec908745bf7f21b22830717517dbf5cd87dda732f7325a8a5f28919530cf7a1b9b2a30de4b3e3a814c1c000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;

exports[`LimitOrderProtocolFacade - facade for Limit order protocol contract fillOrderToWithPermit() Must create a call data for order filling with permit 1`] = `"0xd365c695000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000002a000000000000000000000000000000000000000000000000000000000000003200000000000000000000000000000000000000000000000000de0b6b3a764000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000fb3c7eb936caa12b5a884d612393969a557d430700000000000000000000000000000000000000000000000000000000000003400000000000000000000000000000000000000000000000000000000000000001000000000000000000000000e9e7cea3dedca5984780bafc599bd69add087d56000000000000000000000000111111111117dc0aa78b770fa6a738034120c302000000000000000000000000fb3c7eb936caa12b5a884d612393969a557d4307000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000de0b6b3a76400000000000000000000000000000000000000000000000000000de0b6b3a764000000000024000000240000002400000024000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000140000000000000000000000000000000000000000000000000000000000000002463592c2b00000000000000000000000000000000000000000000000000000000609f1be800000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000004179f0370195da42aef392ddc9d3254827146a8476c559fcec908745bf7f21b22830717517dbf5cd87dda732f7325a8a5f28919530cf7a1b9b2a30de4b3e3a814c1c00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000070234233432032400000000000000000000000000000000000000000000000000"`;

exports[`LimitOrderProtocolFacade - facade for Limit order protocol contract fillRFQOrder() Must create a call data for RFQ order filling 1`] = `"0x3eca9c0a000000000000000000000000000000000000000060bf8c960000000000000002000000000000000000000000ae6c77d06226742a333a6d2991fe3331889c09a600000000000000000000000049feb353fdf1d396a959973216cbac10ef11e7bf00000000000000000000000096577468b160184347e16340a80a9e81c132b9670000000000000000000000009741db81f7b3b23ef66f285ed5c7dc2cb94b601e0000000000000000000000000000000000000000000000000de0b6b3a76400000000000000000000000000000000000000000000000000005a34a38fc00a000000000000000000000000000000000000000000000000000000000000000001208000000000000000000000000000000000000000000000000de0b6b3a76400000000000000000000000000000000000000000000000000000000000000000041dac60ac2a447d98277ca2b39b5fb69274e28851054a4c86cddf7563931d826293a836fb0c3ce60048f267bf9509bd57729a1e8e9fbc91a376ee85016c87beac61c00000000000000000000000000000000000000000000000000000000000000"`;
